### PR TITLE
Newtype for generically permitting thunks in fields

### PR DESCRIPTION
**Pull request by redxaxder**

### CODE REVIEW

**Comment by redxaxder** [comment]
For some reason I was unable to fail the tests locally before, but I now see that this changeset is definitely wrong.

**Comment by redxaxder** [comment]
Well the test passes now! Maybe it's right?

File: src/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs [hunk 1]
```diff
@@ -1,15 +1,18 @@
...
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE KindSignatures        #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE PolyKinds             #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE StandaloneDeriving    #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeOperators         #-}
+{-# LANGUAGE UndecidableInstances  #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
```
Comment (edsko) [inline src/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: Document why this is needed (is it the `HasFields` constraint?).

File: src/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs [hunk 2]
```diff
@@ -106,7 +110,7 @@ class NoUnexpectedThunks a where
   whnfNoUnexpectedThunks :: [String] -> a -> IO ThunkInfo
   default whnfNoUnexpectedThunks :: (Generic a, GWhnfNoUnexpectedThunks (Rep a))
                                  => [String] -> a -> IO ThunkInfo
-  whnfNoUnexpectedThunks ctxt x = gWhnfNoUnexpectedThunks ctxt fp
+  whnfNoUnexpectedThunks ctxt x = gWhnfNoUnexpectedThunks [] ctxt fp
```
Comment (edsko) [inline src/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: I was surprised to see you pass this list at the _value level_, rather than the type level; all metadata is available at the type level, and so all decisions could be made at the type level. But I guess this way works also. Feels slightly less clean, but no big deal.

File: src/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs [hunk 3]
```diff
@@ -296,6 +311,19 @@ instance NoUnexpectedThunks (AllowThunk a) where
...
+instance (HasFields s a, FromSymbols s, Generic a, Typeable a, GWhnfNoUnexpectedThunks (Rep a))
+   => NoUnexpectedThunks (AllowThunksIn s a) where
+  showTypeOf _ = show $ typeRep (Proxy @a)
+  whnfNoUnexpectedThunks ctxt (AllowThunksIn x) = gWhnfNoUnexpectedThunks allowedThunks ctxt fp
+    where
+      fp :: Rep a x
+      !fp = from x
+      allowedThunks = fromSymbols (Proxy @s)
+
+class HasFields (s :: [Symbol]) a
```
Comment (edsko) [inline src/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: The only reason for this class is to make sure the user doesn't have a typo I guess? Maybe document its purpose.

File: test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs [hunk 4]
```diff
@@ -127,8 +128,8 @@ instance FromModel Int where
   fromModel (IntValue n) k = case n of I# result -> k (I# result)
 
   genModel = Gen.choice [
-        IntThunk <$> genModel
-      , IntValue <$> Gen.int Range.linearBounded
+        IntValue <$> Gen.int Range.linearBounded
+      , IntThunk <$> genModel
```
Comment (edsko) [inline test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: Unnecessary change.

File: test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs [hunk 5]
```diff
@@ -152,8 +153,8 @@ instance (FromModel a, FromModel b) => FromModel (a, b) where
                                   k (a', b')
 
   genModel = Gen.choice [
-        PairThunk <$> genModel
-      , PairDefined <$> genModel <*> genModel
+        PairDefined <$> genModel <*> genModel
+      , PairThunk <$> genModel
```
Comment (edsko) [inline test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: Same.

File: test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs [hunk 6]
```diff
@@ -188,8 +189,8 @@ instance FromModel a => FromModel [a] where
       go :: Int -> Gen (Model [a])
       go 0 = pure ListNil
       go n = Gen.choice [
-                 ListThunk <$> go (n - 1)
-               , ListCons <$> genModel <*> go (n - 1)
+                 ListCons <$> genModel <*> go (n - 1)
+               , ListThunk <$> go (n - 1)
```
Comment (edsko) [inline test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: Same.

File: test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs [hunk 7]
```diff
@@ -225,8 +226,8 @@ instance FromModel (Seq Int) where
       -- check) can diverge, because the internal @FingerTree@ may not be
       -- fully evaluated.
       Gen.choice [
-          go genModel sz
-        , go (pure $ IntValue 0) sz
+          go (pure $ IntValue 0) sz
+        , go genModel sz
```
Comment (edsko) [inline test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: Same.

File: test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs [hunk 8]
```diff
@@ -238,6 +239,51 @@ forceSeqToWhnf (SI.Seq SI.EmptyT)          k = k (SI.Seq SI.EmptyT)
...
+
+  modelUnexpected ctxt m =
+    let ctxt' =
+          case modelIsNF ctxt m of
+            IsNF      -> Nothing
+            IsWHNF  a -> Just a
+            NotWHNF a -> Just a
+    in case m of
+      RecordThunk _ -> ctxt'
+      RecordDefined _ (IntThunk _) -> ctxt'
```
Comment (edsko) [inline test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: I think this isn't quite right. You are returning the context computed by `modelIsNF` in the case that the right int is a thunk; but what if the left `Int` is also a thunk? This only works by accident: if the left field would have had a different type, then this would start to fail. So maybe that would a be a good idea (give the two fields different types).

I also find the definition of `modelUnexpected` slightly confusing, because we are doing a repeated pattern match on the same argument `m` within `modelUnexpected`, and moreover call `modelNF` on the exact same value. If we inline some stuff we end up with

```haskell

  modelUnexpected ctxt m =
    case m of
      RecordThunk _ -> Just ctxt' 
      RecordDefined x (IntThunk y) -> mustBeWHNF $ constrNF [modelIsNF ctxt' x, modelIsNF ctxt' (IntThunk y)] of
      RecordDefined x (IntValue y) -> Nothing
    where
      ctxt' = "Record" : ctxt

      mustBeWHNF :: IsNormalForm a -> a
      mustBeWHNF (IsWHNF a) = a
      mustBeWHNF _otherwise = error "mustBeWHNF: expected WHNF"
```

(this is just inlining your definitions), which also makes the bug I mentioned above clearer: we really shouldn't be looking at `x` at all in the `IntThunk` case. Getting rid of looking at `x` gives us

```haskell
  modelUnexpected ctxt m =
    case m of
      RecordThunk _ -> Just ctxt' 
      RecordDefined _ (IntThunk y) -> modelIsNF ctxt' (IntThunk y)
      RecordDefined _ (IntValue y) -> Nothing
    where
      ctxt' = "Record" : ctxt
```

but now we can see that we are duplicating some work that `modelUnexpected` for `Int` is already doing; so this can simplify to just

```haskell
  modelUnexpected ctxt m =
    case m of
      RecordThunk   _   -> Just ctxt' 
      RecordDefined _ y -> modelIsNF' ctxt y
    where
      ctxt' = "Record" : ctxt
```

which is a lot easier to understand: if the think itself is a thunk, this is _definitely_ a thunk; otherwise, we just need to look at the second field.

File: test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs (hunk 4 above)
Comment (redxaxder) [reply test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: The reason for this change (and the similar ones below) is that `Gen.choice` shrinks toward the beginning of the list.

Earlier when I had test failures the shrinker was blowing up because it was building larger and larger trees of `IntThunk`.

So I rearranged all arguments to `Gen.choice` to put the smallest tree at the front.

File: src/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs (hunk 1 above)
Comment (redxaxder) [reply src/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: That one didn't trigger the warning, which seems odd now that you mention it.

File: src/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs (hunk 2 above)
Comment (redxaxder) [reply src/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: Do you have an approach in mind for doing it at the type level?

I'd be up for it.

Comment (redxaxder) [reply src/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: Ah, I think I see it. Let's see how it compares.

File: test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs (hunk 4 above)
Comment (edsko) [reply test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: Aha! Gotcha. Makes perfect sense :) Thanks :)

File: test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs [hunk 9]
```diff
@@ -238,6 +242,47 @@ forceSeqToWhnf (SI.Seq SI.EmptyT)          k = k (SI.Seq SI.EmptyT)
...
+  modelIsNF ctxt = \case
+    RecordThunk _ -> NotWHNF ctxt'
+    RecordDefined a b -> constrNF [modelIsNF ctxt' a, modelIsNF ctxt' b]
+    where
+      ctxt' = "Record" : ctxt
+
+  modelUnexpected ctxt m =
+      case m of
+        RecordThunk   _   -> Just ctxt'
+        RecordDefined _ y -> modelIsNF' ctxt' y
```
Comment (edsko) [inline test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: Why the need to override this? Would the `modelIsNF` function on `Int` not suffice?

File: test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs [hunk 10]
```diff
@@ -238,6 +242,47 @@ forceSeqToWhnf (SI.Seq SI.EmptyT)          k = k (SI.Seq SI.EmptyT)
 forceSeqToWhnf (SI.Seq (SI.Single a))      k = k (SI.Seq (SI.Single a))
 forceSeqToWhnf (SI.Seq (SI.Deep n l ft r)) k = k (SI.Seq (SI.Deep n l ft r))
 
+{-------------------------------------------------------------------------------
+  AllowThunksIn
+-------------------------------------------------------------------------------}
+
+data Record = Record
+  { field1 :: Int
+  , field2 :: Int
```
Comment (edsko) [inline test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: Let's still change one of these two fields to have a different type to catch the bug that I described previously.

File: test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs (hunk 9 above)
Comment (redxaxder) [reply test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: I'm not sure what you're asking. The `modelIsNF'` function here is being called at `Int` here. Are you suggesting we use the default implementation of `modelUnexpected` in terms of `modelIsNF` `@Record`?

My interpretation of the intent behind `modelIsNF` is that it should answer whether the value is in NF or WHNF or not, regardless of whether the deviation from that is allowed. So the implementation here detects the permitted thunk in the first field.

The default implementation of `modelUnexpected` would produce a message for the thunk I wanted to ignore. To make the it do what we want, `modelIsNF` would need to ignore the permitted thunk. Is that fine?

Comment (edsko) [reply test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: I guess what I'm asking is: why not just say

```
  modelUnexpected ctxt m =
      case m of
        RecordThunk   _   -> Just ctxt'
        RecordDefined _ y -> modelUnexpected ctxt y
      where
        ctxt' = "Record" : ctxt
```

saying: "If the record is defined, it has unexpected thunks only if the second field does". Now there is no reason to call that `modelIsNF'` function (or indeed abstract it out at all).

Comment (redxaxder) [reply test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs]: Ah! I like it.
